### PR TITLE
feat: actions

### PR DESCRIPTION
**Pull request by livio-a**

### CODE REVIEW

File: proto/zitadel/action.proto [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            description: "\"the type of the flow\"";
+        }
+    ];
+    zitadel.v1.ObjectDetails details = 2;
+    FlowState state = 3 [
+        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+            description: "the state of the flow";
+        }
+    ];
+    //map<TriggerType, TriggerAction> triggers = 4;
```
Comment (fgerschwiler) [inline proto/zitadel/action.proto:105]: remove if not used

File: internal/query/action.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	From("zitadel.projections.actions").PlaceholderFormat(squirrel.Dollar)
+
+func (q *Queries) GetAction(ctx context.Context, id string, orgID string) (*Action, error) {
+	idQuery, _ := newActionIDSearchQuery(id)
+	actions, err := q.SearchActions(ctx, &ActionSearchQueries{Queries: []SearchQuery{idQuery}})
+	if err != nil {
+		return nil, err
+	}
+	if len(actions) != 1 {
+
```
Comment (fgerschwiler) [inline internal/query/action.go:22]: Nothing happens in this case

File: internal/query/action.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	actions := []*Action{}
+	for rows.Next() {
+		org := new(Action)
+		rows.Scan(
+			&org.CreationDate,
+			&org.ChangeDate,
+			&org.ResourceOwner,
+			//&org.State,
```
Comment (fgerschwiler) [inline internal/query/action.go:45]: remove if not used

File: internal/query/action.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+	return actions, nil
+}
+
+type Action struct {
+	ID            string    `col:"id"`
+	CreationDate  time.Time `col:"creation_date"`
+	ChangeDate    time.Time `col:"change_date"`
+	ResourceOwner string    `col:"resource_owner"`
+	//State         domain.ActionState `col:"action_state"`
```
Comment (fgerschwiler) [inline internal/query/action.go:68]: remove if not used

File: internal/query/action_flow.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+)
+
+func (q *Queries) GetActionsByFlowAndTriggerType(ctx context.Context, flowType domain.FlowType, triggerType domain.TriggerType) ([]*Action, error) {
+	flowTypeQuery, _ := NewTriggerActionFlowTypeSearchQuery(flowType)
+	triggerTypeQuery, _ := NewTriggerActionTriggerTypeSearchQuery(triggerType)
+	return q.SearchActionsFromFlow(ctx, &TriggerActionSearchQueries{Queries: []SearchQuery{flowTypeQuery, triggerTypeQuery}})
+}
+
+var triggerActionsQuery = squirrel.StatementBuilder.Select("creation_date", "change_date", "resource_owner", "sequence", "action_id", "name", "script", "trigger_type", "trigger_sequence").
+	From("zitadel.projections.flows_actions_triggers").PlaceholderFormat(squirrel.Dollar)
```
Comment (fgerschwiler) [inline internal/query/action_flow.go:20]: why is this format dollar?

File: internal/query/action_flow.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	for rows.Next() {
+		action := new(Action)
+		var triggerType domain.TriggerType
+		var triggerSequence int
+		rows.Scan(
+			&action.CreationDate,
+			&action.ChangeDate,
+			&action.ResourceOwner,
+			&action.Sequence,
+			//&action.State,
```
Comment (fgerschwiler) [inline internal/query/action_flow.go:43]: remove if not used

File: internal/query/action_flow.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
```
Comment (fgerschwiler) [inline internal/query/action_flow.go:88]: remove if not used

File: internal/query/projection/flow/flow.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+const (
+	triggerTableSuffix   = "triggers"
+	flowTypeCol          = "flow_type"
+	flowTriggerTypeCol   = "trigger_type"
+	flowChangeDateCol    = "change_date"
+	flowResourceOwnerCol = "resource_owner"
+	flowStateCol         = "flow_state"
+	//flowSequenceCol      = "sequence"
```
Comment (fgerschwiler) [inline internal/query/projection/flow/flow.go:69]: remove if not used

File: internal/query/projection/flow/flow.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	flowActionTriggerSequenceCol = "trigger_sequence"
+	flowActionIDCol              = "action_id"
+
+	actionTableSuffix = "actions"
+	//actionFlowTypeCol      = "flow_type"
```
Comment (fgerschwiler) [inline internal/query/projection/flow/flow.go:74]: remove if not used

File: internal/query/projection/flow/flow.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+			handler.NewCond(flowTypeCol, e.FlowType),
+			handler.NewCond(flowTriggerTypeCol, e.TriggerType),
+		},
+		crdb.WithTableSuffix(triggerTableSuffix),
+	)
+	for i, id := range e.ActionIDs {
+		stmts[i+1] = crdb.AddCreateStatement(
+			[]handler.Column{
+				handler.NewCol(flowResourceOwnerCol, e.Aggregate().ResourceOwner),
+				//handler.NewCol(flowSequenceCol, e.Sequence()),
```
Comment (fgerschwiler) [inline internal/query/projection/flow/flow.go:103]: remove if not used

File: internal/query/projection/projection.go [hunk 11]
```diff
@@ -37,9 +37,11 @@ func Start(ctx context.Context, es *eventstore.Eventstore, config Config) error
 		BulkLimit:         config.BulkLimit,
 	}
 
-	NewOrgProjection(ctx, applyCustomConfig(projectionConfig, config.Customizations["orgs"]))
-	NewProjectProjection(ctx, applyCustomConfig(projectionConfig, config.Customizations["projects"]))
-	owner.NewOrgOwnerProjection(ctx, applyCustomConfig(projectionConfig, config.Customizations["org_owners"]))
+	//NewOrgProjection(ctx, applyCustomConfig(projectionConfig, config.Customizations["orgs"]))
+	//NewProjectProjection(ctx, applyCustomConfig(projectionConfig, config.Customizations["projects"]))
+	//owner.NewOrgOwnerProjection(ctx, applyCustomConfig(projectionConfig, config.Customizations["org_owners"]))
```
Comment (fgerschwiler) [inline internal/query/projection/projection.go:42]: Remove if not used? or is this already on main?

File: internal/query/query.go [hunk 12]
```diff
@@ -22,33 +25,45 @@ type Queries struct {
...
 	Eventstore types.SQLUser
 }
 
 func StartQueries(ctx context.Context, es *eventstore.Eventstore, projections projection.Config, defaults sd.SystemDefaults) (repo *Queries, err error) {
+	sqlClient, err := projections.CRDB.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	sq.StatementBuilder = sq.StatementBuilder.PlaceholderFormat(sq.Dollar) //TODO: ?
```
Comment (fgerschwiler) [inline internal/query/query.go:42]: todo?

File: internal/query/query.go [hunk 13]
```diff
@@ -22,33 +25,45 @@ type Queries struct {
...
+
 	repo = &Queries{
 		iamID:       defaults.IamID,
 		eventstore:  es,
 		idGenerator: id.SonyFlakeGenerator,
+		client:      sqlClient,
+		//querier: querier,
```
Comment (fgerschwiler) [inline internal/query/query.go:49]: remove if not used

File: internal/query/query.go [hunk 14]
```diff
@@ -22,33 +25,45 @@ type Queries struct {
...
 	org.RegisterEventMappers(repo.eventstore)
 	project.RegisterEventMappers(repo.eventstore)
+	action.RegisterEventMappers(repo.eventstore)
 
 	repo.secretCrypto, err = crypto.NewAESCrypto(defaults.IDPConfigVerificationKey)
 	if err != nil {
 		return nil, err
 	}
 
 	// turned off for this release
```
Comment (fgerschwiler) [inline internal/query/query.go:62]: remove comment

File: internal/ui/login/handler/custom_action.go [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+//			ProjectID:      grant.ProjectID,
+//			ProjectGrantID: grant.ProjectGrantID,
+//			RoleKeys:       grant.Roles,
+//		}, request.UserOrgID)
+//		if err != nil {
+//			return err
+//		}
+//	}
+//	return nil
+//}
```
Comment (fgerschwiler) [inline internal/ui/login/handler/custom_action.go:69]: remove if not used

**Review by fgerschwiler — COMMENTED** [review]
Tests on command and query/projection side missing

File: internal/actions/api.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+	a.set("setPhoneVerified", func(verified bool) {
+		if human.Phone == nil {
+			return
+		}
+		human.Phone.IsPhoneVerified = verified
+	})
+	return a
+}
+
+func (a *API) SetExternalUser(user *domain.ExternalUser) *API {
```
[the remark this answers is not in the document]
Comment (livio-a) [reply internal/actions/api.go:63]: this func allows users to overwrite certain properties of the external user by script
i could of course add the id as well, but as of now it is not necessary

File: internal/query/projection/projection.go (hunk 11 above)
Comment (livio-a) [reply internal/query/projection/projection.go:43]: on main query does not start at all... this branch requires query side, but not these projections
I made a comment in the code

File: internal/ui/login/handler/jwt_handler.go [hunk 17]
```diff
@@ -82,45 +83,16 @@ func (l *Login) handleJWTExtraction(w http.ResponseWriter, r *http.Request, auth
...
 	if len(metadata) > 0 {
 		authReq, err = l.authRepo.AuthRequestByID(r.Context(), authReq.ID, authReq.AgentID)
 		if err != nil {
 			l.renderError(w, r, authReq, err)
 			return
 		}
-		_, err = l.command.BulkSetUserMetadata(setContext(r.Context(), authReq.UserOrgID), authReq.UserID, authReq.UserOrgID, externalUser.Metadatas...)
+		_, err = l.command.BulkSetUserMetadata(setContext(r.Context(), authReq.UserOrgID), authReq.UserID, authReq.UserOrgID, metadata...)
 		if err != nil {
 			l.renderError(w, r, authReq, err)
```
Comment (fgerschwiler) [inline internal/ui/login/handler/jwt_handler.go:97]: should we really show error page (dead end)? not a page with the error message?

**Review by fgerschwiler — APPROVED** [review]

**Pull request merged**